casper/src/main/scala/coop/rchain/casper/engine/Running.scala

```diff
@@ -8,84 +8,217 @@ import cats.{Applicative, Monad}
+  ): F[Unit] =
+    for {
+      conf <- RPConfAsk[F].ask
+      msg = packet(
+        conf.local,
+        conf.networkId,
+        transport.BlockRequest,
+        BlockRequest(hash).toByteString
+      )
+      _ <- TransportLayer[F].send(peer, msg)
+    } yield ()
+
+  /**
+    * This method should be called periodically to maintain liveness of the protocol
+    * and keep the requested blocks list clean.
+    * See spec RunningMaintainRequestedBlocksSpec for more details
+    */
+  def maintainRequestedBlocks[F[_]: Monad: RPConfAsk: RequestedBlocks: TransportLayer: Log: Time]
+      : F[Unit] = {
+
+    def toMap(list: List[(BlockHash, Option[Requested])]): Map[BlockHash, Requested] = {
+      val filtered = list.flatMap {
+        case (hash, Some(r)) => List((hash -> r))
+        case (_, None)       => List.empty
+      }
+      Map(filtered: _*)
+    }
+
+    def tryRerequest(hash: BlockHash, requested: Requested): F[(BlockHash, Option[Requested])] =
+      if (requested.waitingList.nonEmpty) {
+        val nextPeer = requested.waitingList(0)
+        def modifiedRequested(ts: Long) = requested.copy(
+          timestamp = ts,
+          waitingList = requested.waitingList.tail,
+          peers = requested.peers + nextPeer
+        )
+        for {
+          _  <- requestForBlock[F](nextPeer, hash)
+          ts <- Time[F].currentMillis
+        } yield ((hash -> Option(modifiedRequested(ts))))
+      } else {
+        val warnMessage = s"Could not retrieve requested block ${PrettyPrinter.buildString(hash)}. " +
+          "Removing the request from the requested blocks list. Casper will have to re-request the block."
+        Log[F].warn(warnMessage).as((hash -> none[Requested]))
+      }
+
+    RequestedBlocks[F].flatModify(requests => {
+      requests.keys.toList
+        .traverse(hash => {
+          val requested = requests(hash)
+          Time[F].currentMillis
+            .map(_ - requested.timestamp > timeout.toMillis)
+            .ifM(tryRerequest(hash, requested), (hash -> Option(requested)).pure[F])
+
+        })
+        .map(list => toMap(list))
+    })
+  }
+
+  def handleHasBlock[F[_]: Monad: RPConfAsk: RequestedBlocks: TransportLayer: Time: Log](
+      peer: PeerNode,
+      hb: HasBlock
+  )(
+      casperContains: BlockHash => F[Boolean]
+  ): F[Unit] = {
+
+    val hashStr = PrettyPrinter.buildString(hb.hash)
+
+    def addNewEntry =
+      Log[F].info(s"Creating new entry for the $hashStr request") >> Time[F].currentMillis >>= (
+          ts => {
+            RequestedBlocks[F].modify(
+              _ + (hb.hash -> Requested(timestamp = ts, peers = Set(peer)))
+            )
+          }
+      )
+
+    def addToWaitingList(requested: Requested): F[Unit] =
+      Log[F].info(s"Adding $peer to waiting list of $hashStr request") >> RequestedBlocks[F]
+        .modify { requestedBlocks =>
+          requestedBlocks + (hb.hash -> requested.copy(waitingList = peer :: requested.waitingList))
```

Comment:
We should add a TODO or a ticket to append last

Reply:
https://rchain.atlassian.net/browse/RCHAIN-3675